snuba/migrations/snuba_migrations/outcomes/0003_outcomes_add_category_and_quantity.py

```diff
@@ -0,0 +1,168 @@
+from typing import Sequence
+
+from snuba.clickhouse.columns import Column, UInt, String, DateTime
+from snuba.clusters.storage_sets import StorageSetKey
+from snuba.migrations import migration, operations
+from snuba.migrations.columns import MigrationModifiers as Modifiers
+from sentry_relay import DataCategory
+
+old_materialized_view_columns: Sequence[Column[Modifiers]] = [
+    Column("org_id", UInt(64)),
+    Column("project_id", UInt(64)),
+    Column("key_id", UInt(64)),
+    Column("timestamp", DateTime()),
+    Column("outcome", UInt(8)),
+    Column("reason", String()),
+    Column("times_seen", UInt(64)),
+]
+
+new_materialized_view_columns: Sequence[Column[Modifiers]] = [
+    Column("org_id", UInt(64)),
+    Column("project_id", UInt(64)),
+    Column("key_id", UInt(64)),
+    Column("timestamp", DateTime()),
+    Column("outcome", UInt(8)),
+    Column("reason", String()),
+    Column("category", UInt(8, Modifiers(nullable=True))),
+    Column("quantity", UInt(64, Modifiers(nullable=True))),
+    Column("times_seen", UInt(64)),
+]
+
+
+class Migration(migration.MultiStepMigration):
+    """
+    Adds quantity and category columns to outcomes. updates materialized view and hourly table to support
+    category as a new dimension and quantity as a new measure.
+    """
+
+    blocking = False
+
+    def forwards_local(self) -> Sequence[operations.Operation]:
+        return [
+            operations.AddColumn(
+                storage_set=StorageSetKey.OUTCOMES,
+                table_name="outcomes_raw_local",
+                column=Column("quantity", UInt(32, Modifiers(nullable=True))),
+                after=None,
+            ),
+            operations.AddColumn(
+                storage_set=StorageSetKey.OUTCOMES,
+                table_name="outcomes_raw_local",
+                column=Column("category", UInt(8, Modifiers(nullable=True))),
+                after=None,
+            ),
+            operations.AddColumn(
+                storage_set=StorageSetKey.OUTCOMES,
+                table_name="outcomes_hourly_local",
+                column=Column("quantity", UInt(64, Modifiers(nullable=True))),
+                after=None,
+            ),
+            operations.RunSql(
+                storage_set=StorageSetKey.OUTCOMES,
+                statement="""
+                    ALTER TABLE outcomes_hourly_local ADD COLUMN IF NOT EXISTS category UInt8,
+                    MODIFY ORDER BY (org_id, project_id, key_id, outcome, reason, timestamp, category);
+                """,
+            ),  # note: this migration is not reversable!
```

Comment:
How come it's not reversible? Can't we first modify the order by to drop the reference to the column, then after that drop the column?

Reply:
You can only update the order by, when adding new column, and only by adding the new column to the end.
https://clickhouse.tech/docs/en/sql-reference/statements/alter/order-by/

Though we may actually decide to rebuild the table to add the category before the timestamp, if we end up doing that, this migration will look very different.

Reply:
good call! i had them in the wrong order when i originally tried that. works locally, will push up.

Reply:
mh, ok ignore my comment, this was not used to work not too long ago.

Reply:
@fpacifici this is what I thought as well, but I don't actually get any errors when trying this way 🤔